Creating run backfill

aritra24 opened the pull request:
Resolves: #43969

<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

aritra24 committed:
WIP: Creating run backfill

bbovenzi commented on airflow/ui/src/components/Menu/RunBackfillForm.tsx:
```diff
@@ -0,0 +1,174 @@
+                  {...field}
+                  min={dataIntervalStart || undefined}
+                  onBlur={resetDateError}
+                  placeholder="yyyy-mm-ddThh:mm"
```
I don't think we need this placeholder for `datetime-local`

aritra24 replied:
Ack

bbovenzi commented on airflow/ui/src/components/Menu/RunBackfillForm.tsx:
```diff
@@ -0,0 +1,174 @@
+import { useForm, Controller } from "react-hook-form";
+
+import type { DAGResponse, DAGWithLatestDagRunsResponse } from "openapi/requests/types.gen";
+import { useTrigger } from "src/queries/useTrigger";
```
We're not triggering a dag run, we should use `useBackfillServiceCreateBackfillDryRun` and `useBackfillServiceCreateBackfill`

aritra24 replied:
Ack

bbovenzi commented on airflow/ui/src/components/Menu/RunBackfillForm.tsx:
```diff
@@ -0,0 +1,174 @@
+  readonly onClose: () => void;
+};
+
+export type DagRunTriggerParams = {
```
Let's import `BackfillPostBody` instead

aritra24 replied:
Ack

bbovenzi commented on airflow/ui/src/components/Menu/RunBackfillForm.tsx:
```diff
@@ -0,0 +1,174 @@
+    setErrors((prev) => ({ ...prev, date: undefined }));
+  };
+
+  const runTypes = [
```
Let's rename this to Reprocess Behavior because `dag_run.run_type` is already a concept.

aritra24 replied:
Ack

bbovenzi commented on airflow/ui/src/components/Menu/RunBackfillForm.tsx:
```diff
@@ -0,0 +1,174 @@
+                  </RadioGroup.Item>
+                ))}
+              </VStack>
+            </RadioGroup.Root>
```
We may want to use the RadioCard component in `components/ui` instead.

aritra24 replied:
Ack

bbovenzi reviewed (changes requested):
Thanks for taking this on!

We need to change the fields from DagRun Trigger to Backfill. Here are the fields we need to have:

```
export type BackfillPostBody = {
  dag_id: string;
  from_date: string;
  to_date: string;
  run_backwards?: boolean;
  dag_run_conf?: {
    [key: string]: unknown;
  };
  reprocess_behavior?: "failed" | "completed" | "none";
  max_active_runs?: number;
};
```

aritra24 committed:
ui changes, some temp code to be filled in

aritra24 committed:
Fixing up a bunch of UI

Starting to call the APIs still has some issues

aritra24 committed:
remove null note in ActionAccordion

bbovenzi commented on airflow/ui/src/queries/useCreateBackfill.ts:
```diff
@@ -0,0 +1,86 @@
+
+  const createBackfill = (data: CreateBackfillData) => {
+    const DataIntervalStart = new Date(data.requestBody.from_date);
+    const DataIntervalEnd = new Date(data.requestBody.to_date);
```
```suggestion
    const dataIntervalStart = new Date(data.requestBody.from_date);
    const dataIntervalEnd = new Date(data.requestBody.to_date);
```

Normal variables start with a lowercase letter

aritra24 replied:
Ack

bbovenzi commented on airflow/ui/src/components/Menu/RunBackfillModal.tsx:
```diff
@@ -0,0 +1,48 @@
+  <Dialog.Root lazyMount onOpenChange={onClose} open={open} size="xl" unmountOnExit>
+    <Dialog.Content backdrop>
+      <Dialog.Header bg="blue.muted">
+        {/* <VStack align="start" gap={4}> */}
```
Delete this line?

aritra24 replied:
Ack

bbovenzi commented on airflow/ui/src/components/Menu/RunBackfillForm.tsx:
```diff
@@ -0,0 +1,226 @@
+    { label: "Missing Runs", value: "failed" },
+    { label: "Missing and Errored Runs", value: "completed" },
+    { label: "All Runs", value: "none" },
+  ];
```
Let's move this outside of the component.

aritra24 replied:
Ack

aritra24 committed:
Updated UI

Added an alert for number of backfills, fixed the dryrun request,
fixed up a bit of the UI.

bbovenzi commented on airflow/ui/src/components/Menu/RunBackfillForm.tsx:
```diff
@@ -0,0 +1,217 @@
+        <Controller
+          control={control}
+          name="run_backwards"
+          render={({ field }) => (
```
Let's make sure this colorPalette is also blue

aritra24 replied:
done

bbovenzi commented on airflow/ui/src/components/Menu/RunBackfillForm.tsx:
```diff
@@ -0,0 +1,217 @@
+          <Button onClick={() => void handleSubmit(onCancel)()}>Cancel</Button>
+          <Button
+            colorPalette="blue"
+            disabled={Boolean(errors.date) || affectedTasks.total_entries === 0}
```
This should be disabled if we're loading the dry run.

aritra24 replied:
done

bbovenzi commented on airflow/ui/src/constants/reprocessBehaviourParams.ts:
```diff
@@ -0,0 +1,23 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+export const ReprocessBehaviors = [
```
```suggestion
export const reprocessBehaviors = [
```

bbovenzi commented on airflow/ui/src/components/Menu/RunBackfillButton.tsx:
```diff
@@ -0,0 +1,44 @@
+
+  return (
+    <Box>
+      <Link aria-label="Run Backfill" onClick={onOpen} to="">
```
Let's make this a Button instead.

aritra24 replied:
Ack

aritra24 committed:
Resolved review comments

bbovenzi reviewed (approved):
nice work. Thanks for putting up with all my comments

pierrejeambrun commented on airflow/ui/src/components/ActionAccordion/ActionAccordion.tsx:
```diff
@@ -29,13 +29,14 @@ import { columns } from "./columns";
 // TODO: Make a front-end only unconnected table component with client side ordering and pagination
 const ActionAccordion = ({ affectedTasks, note, setNote }: Props) => {
   const showTaskSection = affectedTasks !== undefined;
+  const showNoteSection = note !== null;
```
What was the motivation for this ?

It looks like the Note section is now missing when we are Clearing or MarkingAs in case we want to add a new note. We can basically only 'edit' an existing note but not add one.

![Screenshot 2025-02-24 at 17 55 04](https://github.com/user-attachments/assets/e4eb8e4f-9667-4e78-815c-4a80681bdae2)

aritra24 replied:
Hey @pierrejeambrun, a miss from me, while I was trying to use this to display backfills I thought the add not was happening at a different place so thought if there were no notes we wouldn't want to show an empty section. I can revert it back since I ended up not using the accordion in my dialog box. Thanks for pointing it out.

pierrejeambrun replied:
It would be great to make a PR that reverts this part back to the initial behavior. 👍

aritra24 replied:
Yup, I have the code pushed with the revert, but I'm not at my laptop right now. I'll raise a pr once I get home in a couple of hours?

pierrejeambrun replied:
Of course, no worry thanks

aritra24 replied:
Raised: https://github.com/apache/airflow/pull/47054
